flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecWindowRank.java

```diff
@@ -0,0 +1,235 @@
+        this.windowing = windowing;
+    }
+
+    @JsonCreator
+    public StreamExecWindowRank(
+            @JsonProperty(FIELD_NAME_RANK_TYPE) RankType rankType,
+            @JsonProperty(FIELD_NAME_PARTITION_SPEC) PartitionSpec partitionSpec,
+            @JsonProperty(FIELD_NAME_SORT_SPEC) SortSpec sortSpec,
+            @JsonProperty(FIELD_NAME_RANK_RANG) RankRange rankRange,
+            @JsonProperty(FIELD_NAME_OUTPUT_RANK_NUMBER) boolean outputRankNumber,
+            @JsonProperty(FIELD_NAME_WINDOWING) WindowingStrategy windowing,
+            @JsonProperty(FIELD_NAME_ID) int id,
+            @JsonProperty(FIELD_NAME_INPUT_PROPERTIES) List<InputProperty> inputProperties,
+            @JsonProperty(FIELD_NAME_OUTPUT_TYPE) RowType outputType,
+            @JsonProperty(FIELD_NAME_DESCRIPTION) String description) {
+        super(id, inputProperties, outputType, description);
+        checkArgument(inputProperties.size() == 1);
+        this.rankType = checkNotNull(rankType);
+        this.partitionSpec = checkNotNull(partitionSpec);
+        this.sortSpec = checkNotNull(sortSpec);
+        this.rankRange = checkNotNull(rankRange);
+        this.outputRankNumber = outputRankNumber;
+        this.windowing = checkNotNull(windowing);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Transformation<RowData> translateToPlanInternal(PlannerBase planner) {
+        // validate rank type
+        switch (rankType) {
+            case ROW_NUMBER:
+                break;
+            case RANK:
+                throw new TableException("RANK() on window rank is not supported currently.");
+            case DENSE_RANK:
+                throw new TableException("DENSE_RANK() on window rank is not supported currently.");
+            default:
+                throw new TableException(
+                        String.format(
+                                "Rank function %s on window rank is not supported currently.",
+                                rankType));
+        }
+
+        // validate rank range
+        ConstantRankRange constantRankRange;
+        if (rankRange instanceof ConstantRankRange) {
+            constantRankRange = (ConstantRankRange) rankRange;
+        } else {
+            throw new TableException(
+                    String.format(
+                            "Rank strategy %s on window rank is not supported currently.",
+                            rankRange));
+        }
+
+        // validate window strategy
+        if (!windowing.isRowtime()) {
+            throw new TableException("PROC-TIME on window rank is not supported currently.");
+        }
+        int windowEndIndex;
+        if (windowing instanceof WindowAttachedWindowingStrategy) {
+            windowEndIndex = ((WindowAttachedWindowingStrategy) windowing).getWindowEnd();
+        } else {
+            throw new UnsupportedOperationException(
+                    windowing.getClass().getName() + " is not supported yet.");
```

Comment:
Do we have tests for this ?

Reply:
Not yet, there is no possible to go into this branch currently. Because only `StreamPhysicalWindowRankRule` could translate `FlinkLogicalRank` to `StreamPhysicalWindowRank` with `WindowAttachedWindowingStrategy`.